### ISSUE TITLE
INTERNAL ERROR: missing context with id = 2

### ISSUE THREAD

**Issue by phawxby** [puppeteer/puppeteer#2811]
TLDR. I'm desperately hoping someone can help with this. I've been at it for 2 days and it's driving me crazy. I'm working on creating a simplified repro case but I'm hoping someone might be able to point me in the right direction. It feels like puppeteer is possibly getting a bit confused about context when running in a highly parallel environment, but I'm probably wrong. 

This problem occurs both locally running Docker and on CircleCI.

### Steps to reproduce

```bash
 FAIL  __tests__/visual-regression/cta.test.js (79.643s)
  ● 99-visual-regression-cta-cta - mobile - ltr

    INTERNAL ERROR: missing context with id = 2

      at assert (node_modules/puppeteer/lib/helper.js:251:11)
      at FrameManager.createJSHandle (node_modules/puppeteer/lib/FrameManager.js:223:5)
      at ExecutionContext.evaluateHandle (node_modules/puppeteer/lib/ExecutionContext.js:89:17)
```

We an enormous visual regression test suite, a hundred or so prebuilt html files each containing a html component. We test each in 3 different viewports both right to left and left to right, so 6 visual regression tests per file using [jest-image-snapshot](https://github.com/americanexpress/jest-image-snapshot). We have the tests broken down into a bunch of different files so jest is able to run them in parallel. 

**Tell us about your environment:**

* Puppeteer version: 1.5.1
* Platform / OS version: [circleci/node:8-browsers](https://hub.docker.com/r/circleci/node/)
* Node.js version: 8
* Jest: 22.4.3
* Jest-image-snapshot: 2.4.0

**Sample**

Example runner
```javascript
require('../_lib/visual-regression').performVisualRegression('foo/bar/');
```

Simplified visual-regression.js.
```javascript
function performVisualRegression(dir, viewports = viewportSpecs.map((v) => v.id)) {
  const slugs = getSlugs(dir);

  for (const slug of slugs) {
    for (const viewport of viewportSpecs.filter((v) => viewports.includes(v.id))) {
      for (const direction of ['ltr', 'rtl']) {
        // test.concurrent
        test(`${slug} - ${viewport.id} - ${direction}`, performTest(slug, viewport, customSnapshotsDir, direction), 30000);
      }
    }
  }
}

function performTest(slug, viewport, customSnapshotsDir, direction = 'ltr') {
  const url = `${global.BASE_URL}/patterns/${slug}/${slug}.rendered.html`;

  return async() => {
    let page;

    try {
      page = await browser.newPage();
      await page.setCacheEnabled(false);
      await page.setViewport(viewport);

      const response = await page.goto(url, { timeout: 15000, waitUntil: ['load', 'networkidle2'] });
      expect(await response.ok()).toBe(true);

      await page.$eval('html', (div, dir) => div.setAttribute('dir', dir), direction);

      const elm = await page.$('body > *:not(script):not(style):not(link):not(meta)');
      expect(elm).not.toBeNull();

      await page.waitFor(250);

      const screenshot = await elm.screenshot();

      expect(screenshot).toBeTruthy();
      expect(screenshot instanceof Buffer).toBe(true);

      let snapshotId = `${name}~~${viewport.id}~~${direction}`;

      const config = Object.assign({}, defaultSnapshotConfig, { customSnapshotsDir, customSnapshotIdentifier: snapshotId });
      expect(screenshot).toMatchImageSnapshot(config);

    } finally {
      if (page) {
        await page.close();
      }
    }
  };
}
```

Jest environment.
```javascript
const NodeEnvironment = require('jest-environment-node');
const puppeteer = require('puppeteer');

class PuppeteerEnvironment extends NodeEnvironment {
  constructor(config) {
    super(config);

    this.browser = puppeteer.launch({
      headless: true,
      dumpio: false,
      args: [
        '--no-sandbox',
        '--disable-setuid-sandbox',
        '--disable-dev-shm-usage',
      ]
    });
  }

  async setup() {
    await super.setup();

    const browser = await this.browser;

    browser.on('targetcreated', async(target) => {
      if (target.type() !== 'page') {
        return;
      }

      const page = await target.page();

      page.evaluateOnNewDocument(() => {
        window.env = 'test';
      });

      page.on('domcontentloaded', async() => {
        await page.evaluate(async() => {
          window.env = 'test';

          // Stop videos from mucking up visual regression tests
          document.querySelectorAll('video').forEach((v) => {
            v.removeAttribute('autoplay');
            v.setAttribute('preload', 'auto');

            v.pause();
            v.currentTime = 0;
            v.load();
          });
        });
      });
    });

    this.global.BROWSER = browser;
  }

  async teardown() {
    await super.teardown();

    const b = await this.browser;
    await b.close();
    this.browser = null;
  }

  dispose() {

  }

  runScript(script) {
    return super.runScript(script);
  }
}

module.exports = PuppeteerEnvironment;
```

I should add we've also been having other extremely weird stuff going on. Left is the reference, middle is the diff, and right is what was produced during the test run.

Viewport not being honoured, completely the wrong size.
![image](https://user-images.githubusercontent.com/1090602/42004340-5bd2b276-7a67-11e8-8cee-6ae49476d4c9.png)

Completely the wrong component
![image](https://user-images.githubusercontent.com/1090602/42004351-6ddf15e0-7a67-11e8-94e8-ca955f987e86.png)

**Comment by aslushnikov** [comment]
Hey @phawxby,

This certainly is our bug. This error can happen if there's a race condition between certain devtools protocol events; however, I don't yet understand where it might come from.

Any repro that I can play with will help a lot. You pasted a lot of code; is this runnable? If yes, how do I run it?

**Comment by phawxby** [comment]
@aslushnikov still working on a repro case but I have narrowed it down to this, removing it removes the error.

```
page.on('domcontentloaded', async() => {
        await page.evaluate(async() => {
          window.env = 'test';

          // Stop videos from mucking up visual regression tests
          document.querySelectorAll('video').forEach((v) => {
            v.removeAttribute('autoplay');
            v.setAttribute('preload', 'auto');

            v.pause();
            v.currentTime = 0;
            v.load();
          });
        });
```

The code above probably isn't runnable, I'll get something runnable to you next week.

**Comment by phawxby** [comment]
@aslushnikov I have managed to recreate the issue with this.
[puppeteer.zip](https://github.com/GoogleChrome/puppeteer/files/2154406/puppeteer.zip)

```
PS C:\Users\Z269CJBB\Documents\GitHub\puppeteer-parallel> yarn run test
yarn run v1.7.0
$ jest
Determining test suites to run...setup

Server running on 5676...






 PASS  __tests__/visual-regression/page-10.test.js (16.4s)
  √ http://localhost:5676/page-10.html - desktop - ltr (2906ms)
  √ http://localhost:5676/page-10.html - desktop - rtl (2537ms)
  √ http://localhost:5676/page-10.html - mobile - ltr (1361ms)
  √ http://localhost:5676/page-10.html - mobile - rtl (1424ms)
  √ http://localhost:5676/page-10.html - small-laptop - ltr (1618ms)
  √ http://localhost:5676/page-10.html - small-laptop - rtl (1612ms)

 PASS  __tests__/visual-regression/page-1.test.js (16.907s)
  √ http://localhost:5676/page-1.html - desktop - ltr (3039ms)
  √ http://localhost:5676/page-1.html - desktop - rtl (2691ms)
  √ http://localhost:5676/page-1.html - mobile - ltr (1389ms)
  √ http://localhost:5676/page-1.html - mobile - rtl (1339ms)
  √ http://localhost:5676/page-1.html - small-laptop - ltr (1718ms)
  √ http://localhost:5676/page-1.html - small-laptop - rtl (1788ms)

 PASS  __tests__/visual-regression/page-7.test.js (17.126s)
  √ http://localhost:5676/page-7.html - desktop - ltr (3126ms)
  √ http://localhost:5676/page-7.html - desktop - rtl (2735ms)
  √ http://localhost:5676/page-7.html - mobile - ltr (1407ms)
  √ http://localhost:5676/page-7.html - mobile - rtl (1377ms)
  √ http://localhost:5676/page-7.html - small-laptop - ltr (1701ms)
  √ http://localhost:5676/page-7.html - small-laptop - rtl (1837ms)

 PASS  __tests__/visual-regression/page-6.test.js (10.73s)
  √ http://localhost:5676/page-6.html - desktop - ltr (2291ms)
  √ http://localhost:5676/page-6.html - desktop - rtl (1996ms)
  √ http://localhost:5676/page-6.html - mobile - ltr (1367ms)
  √ http://localhost:5676/page-6.html - mobile - rtl (1406ms)
  √ http://localhost:5676/page-6.html - small-laptop - ltr (1589ms)
  √ http://localhost:5676/page-6.html - small-laptop - rtl (1591ms)

 PASS  __tests__/visual-regression/page-9.test.js (11.369s)
  √ http://localhost:5676/page-9.html - desktop - ltr (2338ms)
  √ http://localhost:5676/page-9.html - desktop - rtl (2169ms)
  √ http://localhost:5676/page-9.html - mobile - ltr (1361ms)
  √ http://localhost:5676/page-9.html - mobile - rtl (1440ms)
  √ http://localhost:5676/page-9.html - small-laptop - ltr (1604ms)
  √ http://localhost:5676/page-9.html - small-laptop - rtl (1951ms)

 PASS  __tests__/visual-regression/page-8.test.js (11.358s)
  √ http://localhost:5676/page-8.html - desktop - ltr (2329ms)
  √ http://localhost:5676/page-8.html - desktop - rtl (2176ms)
  √ http://localhost:5676/page-8.html - mobile - ltr (1378ms)
  √ http://localhost:5676/page-8.html - mobile - rtl (1374ms)
  √ http://localhost:5676/page-8.html - small-laptop - ltr (1670ms)
  √ http://localhost:5676/page-8.html - small-laptop - rtl (1935ms)

 PASS  __tests__/visual-regression/page-5.test.js (11.641s)
  √ http://localhost:5676/page-5.html - desktop - ltr (2713ms)
  √ http://localhost:5676/page-5.html - desktop - rtl (2491ms)
  √ http://localhost:5676/page-5.html - mobile - ltr (1372ms)
  √ http://localhost:5676/page-5.html - mobile - rtl (1378ms)
  √ http://localhost:5676/page-5.html - small-laptop - ltr (1605ms)
  √ http://localhost:5676/page-5.html - small-laptop - rtl (1585ms)

 PASS  __tests__/visual-regression/page-4.test.js (12.309s)
  √ http://localhost:5676/page-4.html - desktop - ltr (2849ms)
  √ http://localhost:5676/page-4.html - desktop - rtl (2311ms)
  √ http://localhost:5676/page-4.html - mobile - ltr (1471ms)
  √ http://localhost:5676/page-4.html - mobile - rtl (1402ms)
  √ http://localhost:5676/page-4.html - small-laptop - ltr (1766ms)
  √ http://localhost:5676/page-4.html - small-laptop - rtl (1710ms)

 PASS  __tests__/visual-regression/page-2.test.js (12.245s)
  √ http://localhost:5676/page-2.html - desktop - ltr (2701ms)
  √ http://localhost:5676/page-2.html - desktop - rtl (2330ms)
  √ http://localhost:5676/page-2.html - mobile - ltr (1435ms)
  √ http://localhost:5676/page-2.html - mobile - rtl (1448ms)
  √ http://localhost:5676/page-2.html - small-laptop - ltr (1801ms)
  √ http://localhost:5676/page-2.html - small-laptop - rtl (1778ms)

 FAIL  __tests__/visual-regression/page-3.test.js (9.645s)
  √ http://localhost:5676/page-3.html - desktop - ltr (2434ms)
  √ http://localhost:5676/page-3.html - desktop - rtl (2017ms)
  √ http://localhost:5676/page-3.html - mobile - ltr (1341ms)
  × http://localhost:5676/page-3.html - mobile - rtl (142ms)
  √ http://localhost:5676/page-3.html - small-laptop - ltr (1739ms)
  √ http://localhost:5676/page-3.html - small-laptop - rtl (1632ms)

  ● http://localhost:5676/page-3.html - mobile - rtl

    INTERNAL ERROR: missing context with id = 2

      at assert (node_modules/puppeteer/lib/helper.js:251:11)
      at FrameManager.createJSHandle (node_modules/puppeteer/lib/FrameManager.js:223:5)
      at ExecutionContext.evaluateHandle (node_modules/puppeteer/lib/ExecutionContext.js:89:17)

Test Suites: 1 failed, 9 passed, 10 total
Tests:       1 failed, 59 passed, 60 total
Snapshots:   0 total
Time:        52.761s
Ran all test suites.
teardown
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by JoelEinbinder** [comment]
I can't repro the error with the test suite you sent. But I think I know what is going on.

In your setup script, you run `page.evaluate` on `domcontentloaded`. Then when your test is done, you close the page. But you don't await the `page.evaluate` before calling `page.close`. So the page closes before the evaluation finishes, and when the evaluation does finish, the page has already closed and we fail to return the result of the evaluation with this error.

**Comment by aslushnikov** [comment]
@phawxby Joel's explanation is very plausible. I'd catch the error in the `page.evaluate` command to workaround this race:

```js
page.on('domcontentloaded', () => {
  page.evaluate(() => {
    window.env = 'test';

    // Stop videos from mucking up visual regression tests
    document.querySelectorAll('video').forEach((v) => {
      v.removeAttribute('autoplay');
      v.setAttribute('preload', 'auto');
 
      v.pause();
      v.currentTime = 0;
      v.load();
    });
  }).catch(e => {});
});
```

**Comment by phawxby** [comment]
That makes sense. Do you think it would make sense that `page.close()` should not actually close the page until everything has finished evaluating to avoid this error? I think that would mean tracking everything in an array at the page level then just doing an `await Promise.all(commands)` before continuing with the close.

**Comment by aslushnikov** [comment]
>  Do you think it would make sense that page.close() should not actually close the page until everything has finished evaluating to avoid this error?

I don't think so; it's up to a client to await all pending operations or interrupt them by closing the page.

**Comment by JoelEinbinder** [comment]
It is pretty much always wrong to not await code being run on events. You have another race in this code, which might no be manifesting. On `domcontentloaded` you set `window.env = 'test';`. But you don't wait for this eval to run before moving on with your test. An indefinite amount of time could pass between the `domcontentloaded` event firing in Chromium, the `domcontentloaded` event firing in node, and then running the code in Chromium again via your evaluate. In your setup function, it should look something like this:
```js
await new Promise(x => page.once('domcontentloaded', x);
await page.evaluate(() => ...);
```
